src/SwiftPackage.ts

```diff
@@ -45,39 +45,55 @@ export interface Dependency {
     url?: string
 }
 
+// package we attempted to load but failed
+class InvalidPackage implements PackageContents {
+    get name(): string { return ""; }
+    get products(): Product[] { return []; }
+    get dependencies(): Dependency[] { return []; }
+    get targets(): Target[] { return []; }
+}
+
 // Class holding Swift Package Manager Package
 export class SwiftPackage implements PackageContents {
 	private constructor(
         readonly folder: string,
-        public contents?: PackageContents
+        public contents?: PackageContents|null
```

Comment:
Do you just want this to be an optional? If so, it should just be `contents?: PackageContents`. The type will then be `PackageContents | undefined`. Currently, it's `PackageContents | undefined | null`. Or do you want to treat `null` differently from `undefined`?

Reply:
I struggled to return a `Promise<undefined>` from `loadPackage` so just switched to `null`

Reply:
Perhaps we could just use `PackageContents | null | undefined` to differentiate between the three states (valid / invalid / no package.swift)?
The getters already return the default values, so it feels a bit redundant to add an `InvalidPackage` class just to indicate a state.

Something like this:

```typescript
export class SwiftPackage implements PackageContents {
    private constructor(
        readonly folder: string,
        public contents: PackageContents | null | undefined
    ) {
        this.setContextKeys();
    }

    public static async create(folder: string): Promise<SwiftPackage> {
        let contents = await SwiftPackage.loadPackage(folder);
        return new SwiftPackage(folder, contents);
    }

    public static async loadPackage(folder: string): Promise<PackageContents | null | undefined> {
        try {
            const { stdout } = await exec('swift package describe --type json', { cwd: folder });
            return JSON.parse(stdout);
        } catch(error) {
            const execError = error as {stderr: string};
            if (execError.stderr.startsWith("error: root manifest")) {
                return undefined;
            } else {
                return null;
            }
        }
    }

    // ...

    // Return if Package.swift is valid
    public get isValid(): boolean {
        return this.contents ? true : false;
    }

    // Did we find a Package.swift
    public get foundPackage(): boolean {
        return this.contents !== undefined;
    }

   // ...

    private setContextKeys() {
        if (!this.contents) {
            contextKeys.hasPackage = false;
            contextKeys.packageHasDependencies = false;
        }
        contextKeys.hasPackage = true;
        contextKeys.packageHasDependencies = this.dependencies.length > 0;  
    }
}
```

Reply:
I did think about using `null` for a package that failed to load but initially felt it wasn't explicit enough in indicating a package failed to load. That's why I came up with the new type but given this is all internal I guess it wouldn't hurt just using `null`

Reply:
In Swift we'd probably use an enum so I'd make this as explicit as possible

Reply:
Unfortunately, enums in TypeScript aren't as powerful as they are in Swift, so we'd have to add an enum field just for the state, and store the actual state separately in `contents`. But that means we could have an inconsistent state where, for example, `state === .valid` but `contents === undefined`.